### ISSUE TITLE
关于compute_hessian函数的问题

### ISSUE THREAD

**Issue by E18301194** [lorenmt/auto-lambda#4]
您好  我拜读了你的论文 非常感谢您对多任务的贡献  在此有几个问题希望得到您的解答
在代码auto_lambda.py文件夹 compute_hessian函数中，1.  首先对 p += eps * d  后求导self.meta_weights权重，2. 然后 p -= 2 * eps * d，后再一次求导self.meta_weights，3.  最后p += eps * d  计算得到hessian = [(p - n) / (2. * eps) for p, n in zip(d_weight_p, d_weight_n)]
（1）我不明白  p先加上eps * d 再减去2*eps * d，再加上eps * d 是不是相当于p没有变化？ 
（2） d_model  是经过最重要的val_loss更新后的网络权重，我不明白p += eps * d这样做的意义？
（3）因为我这块方向了解不深入，compute_hessian函数应该是核心算法，但是代码这段代码作用我没看明白，希望得到您的指导 万分感谢

**Comment by lorenmt** [comment]
(1) 是的. 我们用p+-eps 来approximate 二阶导, 计算完之后需要恢复参数原本的数值,因此重新加上eps 换回原值.
(2) d_model 是根据 val_loss 计算的一节梯度, 不是网络参数: https://github.com/lorenmt/auto-lambda/blob/24591b7ff0d4498b18bd4b4c85c41864bdfc800a/auto_lambda.py#L66
(3) 你要先理解 AutoLambda的设计哲学: 更新lambda 使得 lambda weighted loss 能帮助取的更好的primary task performance.因为这里是二阶导--> 然后我们用+-eps 来近似二阶导.然后再去看代码怎么写的.

**Comment by lorenmt** [comment]
No further updates.

**Comment by E18301194** [comment]
您好  我仔细阅读了auto lambda论文，这真是一份出色的工作。在您论文中，这部分我没有看懂理解是什么意思：

Note that, λ is only applied on the training loss not validation loss, otherwise, we would easily reach trivial solutions λ = 0. In addition, assuming θ'= θ∗ is also not applicable, otherwise we have ∇λ = 0.

希望得到您的回复 万分感谢

**Comment by lorenmt** [comment]
Hello, what I meant is say you have lambda on validation loss, i.e. min_lambda \sum_i lambda_i \times L_i, then the optimal solution is lambda_i=0 then we have zero loss, which reaches trivial solutions.

Another common assumption is to assume theta' is optimal theta (You can read the DARTS paper, they have the experiments on that assumption), but that also does not apply in our setting: because in that case, we don't have second-order gradients, and directly optimise lambda on Eq 3 (outer loop) would give you zero gradients, because lambdas are not in the loss function.
